### ISSUE TITLE
Image retrain - classification module gets better accuracy than feature_vector

### ISSUE THREAD

**Issue by haoxi911** [tensorflow/hub#43]
I was retraining MobileNet_v1_1.0_224 for image classification using the retrain script in this repo, I was using the same dataset (LSVRC 2012). I noticed that the final test accuracy of feature_vector module was around 55%, while the result of classification module can reach 70%.

I was using same parameters as below: 

--learning_rate=0.005
--testing_percentage=5
--validation_percentage=20
--flip_left_right True
--random_scale=30
--random_brightness=30
--how_many_training_steps=10000

I wonder to know why there is such a huge difference between the two modules, also I supposed we should use the feature_vector one by reading related documents, and in the examples (in comments) in retrain script, we also used feature_vector module by default. 

Can someone explain the reasons of this difference? Is it a problem of `bottleneck` since I distorted images?

**Comment by arnoegw** [comment]
Please clarify: You were using retrain.py with either of
https://tfhub.dev/google/imagenet/mobilenet_v1_100_224/classification/1
https://tfhub.dev/google/imagenet/mobilenet_v1_100_224/feature_vector/1
to train a classifier for the 1000 classes of ILSVRC 2012, that is, the same dataset those modules were trained on?

**Comment by haoxi911** [comment]
@arnoegw  Yes, same dataset and also same parameters on the retrain.py script.

**Comment by arnoegw** [comment]
Which same dataset? The 1 million images of ILSVRC-2012-CLS ("Imagenet")?  That is certainly beyond what the retrain.py example was made for....

**Comment by haoxi911** [comment]
@arnoegw I think so, but I am currently using the retrain.py script to train the model and it seems work (at least by using the `classification` URL.

Would you please let me know if there is a drawback in my case? Thanks!

**Comment by arnoegw** [comment]
Hi haoxi911

If your goal is to do actual ImageNet classification with TensorFlow Hub, I would suggest to use any of the tfhub.dev/google/imagenet/*/classification modules directly, that is: feed their output logits directly into a softmax. For mobilenet_v1_100_224, this should yield the 70.9% top-1 accuracy reported on
https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet_v1.md

Adding a linear transformation in between (like retrain.py does) is not necessary in this case. It only complicates the model, and takes time to train.

retrain.py is meant as a quick way to adapt a module from TensorFlow Hub to a new, small classification problem, using a single machine, possibly without GPU. If I understand your flag settings correctly, you ran it over ImageNet for about one epoch (10000 steps with batch size 100), and probably that took a long time already. But **a single epoch is not a lot to train the final layer carefully** (and certainly doesn't let you benefit from data augmentation). Also, retrain.py does not use the same sophisticated training techniques like full-blown ImageNet training, like learning rate decay, and exponential moving averages of weights (not to mention the weaker regularization).

I am not surprised that tfhub.dev/google/imagenet/mobilenet_v1_100_224/feature_vector/1 falls short of the published accuracy for Mobilenet (70.9%, vs your 55%), given those aspects. I am more surprised that tfhub.dev/google/imagenet/mobilenet_v1_100_224/classification/1 works so much better. Perhaps recovering the identity transformation is easier?!  I have no good intuition for that; the sizes of the coefficient matrices seem comparable (1000x1024 from features to logits, 1000x1000 from logits to logits).

**Comment by haoxi911** [comment]
Hello @arnoegw 

I am so sorry for such a long delay on responding to your comment, as I am still learning the training skills mentioned in your comment, now I think I've made a few progress.

Just to clarify, we are not going to retrain the LSVRC dataset, we are trying to only include a subset of it and combine it with some other images from original ImageNet dataset - this is still in progress, and the overall size of our dataset is comparable to LSVRC.

I made a few changes to the original retrain script, for example, I added `tf.train.exponential_decay` for learning rate decay, also I replaced `GradientDescentOptimizer` with `AdamOptimizer` since the later one has a builtin moving average strategy - though may have a difference with typical EMA.

I am looking into optimizing the regularization, and it seems we can tweak loss by l2_loss and some other variables / hyperparameters, I am wondering if you can provide any insights on this part, and / or if you have any additional suggestions to me.

PS: it may take about ~7 hours to run one epoch, so 10 epochs would be 3 days, I can give it a try during the coming weekend.

Thanks!

**Comment by haoxi911** [comment]
Close this ticket. 

One shall not retrain a model using the `classification` module, using `feature_vector` and append a classifier at the end of the graph and retrain the classifier should be the correct way.
